### PR TITLE
feat: caching the govaluate expressions

### PR DESCRIPTION
**Pull request by timconinx**
Trying to provide a solution for issue https://github.com/casbin/casbin/issues/1028 

All tests pass, but I'm not quite sure if every edge case is covered

Fix: https://github.com/casbin/casbin/issues/1028

### CODE REVIEW

File: enforcer.go [hunk 1]
```diff
@@ -38,10 +39,11 @@ type Enforcer struct {
...
 
-	adapter    persist.Adapter
-	watcher    persist.Watcher
-	dispatcher persist.Dispatcher
-	rmMap      map[string]rbac.RoleManager
+	adapter     persist.Adapter
+	watcher     persist.Watcher
+	dispatcher  persist.Dispatcher
+	rmMap       map[string]rbac.RoleManager
+	expressions sync.Map
```
Comment (tangyang9464) [inline enforcer.go:46]: We should declare ``expressions`` as a variable, not a member of the enforcer structure

File: enforcer.go [hunk 2]
```diff
@@ -537,13 +545,24 @@ func (e *Enforcer) enforce(matcher string, explains *[]string, rvals ...interfac
...
+		if !isPresent {
+			expression, err = govaluate.NewEvaluableExpressionWithFunctions(expString, functions)
+			if err != nil {
+				return false, err
+			}
+			e.expressions.Store(expString, expression)
+		} else {
+			expression = cachedExpression.(*govaluate.EvaluableExpression)
+		}
+	} else {
```
Comment (tangyang9464) [inline enforcer.go:559]: I think the logic of this part of the code should be
1. If hasEval then generateEval function
2.if !Present then new expression
else use cache. 
If hasEval, it may still use cache.

**Review by tangyang9464 — CHANGES_REQUESTED** [review]

File: enforcer.go (hunk 1 above)
Comment (hsluoyz) [inline enforcer.go:46]: rename to `matcherMap`

Comment (hsluoyz) [reply enforcer.go:46]: @tangyang9464 you mean global var? Why not an enforcer member?

File: enforcer.go [hunk 3]
```diff
@@ -471,6 +475,10 @@ func NewEnforceContext(suffix string) EnforceContext {
 	}
 }
 
+func (e *Enforcer) invalidateCache() {
```
Comment (hsluoyz) [inline enforcer.go:478]: invalidateMatcherMap

File: enforcer.go [hunk 4]
```diff
@@ -537,13 +545,24 @@ func (e *Enforcer) enforce(matcher string, explains *[]string, rvals ...interfac
 	var expression *govaluate.EvaluableExpression
 	hasEval := util.HasEval(expString)
 
-	if hasEval {
+	if !hasEval {
+		var cachedExpression, isPresent = e.expressions.Load(expString)
```
Comment (hsluoyz) [inline enforcer.go:549]: This code logic is way too complicated when messed up together, can we extract a wrapper out which will use cache?

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: enforcer.go (hunk 4 above)
Comment (timconinx) [reply enforcer.go]: @hsluoyz do you mean wrap the caching code in a separate func?

Comment (timconinx) [reply enforcer.go]: Code cleanup done, and made the logic easier to understand

File: enforcer.go (hunk 2 above)
Comment (timconinx) [reply enforcer.go]: When using the cache with 'hasEval', a lot of tests fail... For now I'm restricting the cache to matchers without eval

File: enforcer.go (hunk 1 above)
Comment (tangyang9464) [reply enforcer.go]: I made a mistake. It's ok.

File: enforcer.go (hunk 2 above)
Comment (tangyang9464) [reply enforcer.go]: That's weird. Did you find out why?

Comment (timconinx) [reply enforcer.go]: At first glance it looks like the cached key is the matcher-string with the eval not filled in, while the EvaluableExpression has the request parameters filled in. This means that a second call, with different request parameters gets a hit on the matcher string, but uses the Expression with the parameters of the previous request.
I'm not sure how to solve this, but I'm inclined to believe that eval matchers are "worth" to be regenerated into new Expressions each time, because of the added eval-complexity.

**Comment by abichinger** [comment]
@timconinx Please make sure, that the PR title and commit messages follow the [Conventional Commits spec](https://www.conventionalcommits.org/en/v1.0.0/#summary)

Otherwise it looks good to me.

File: util/builtin_operators.go [hunk 5]
```diff
@@ -331,7 +332,7 @@ func GlobMatchFunc(args ...interface{}) (interface{}, error) {
 
 // GenerateGFunction is the factory method of the g(_, _[, _]) function.
 func GenerateGFunction(rm rbac.RoleManager) govaluate.ExpressionFunction {
-	memorized := map[string]bool{}
+	memorized := sync.Map{}
```
Comment (thetorpedodog) [inline util/builtin_operators.go:335]: This was something that got me when I worked in this function earlier—at the time, the `g` function was called by govaluate purely serially, within the context of exactly one enforcement call. Is this something that is changed by the new expression caching?

**Review by thetorpedodog — COMMENTED** [review]
Hello! I recently did some work in this corner of the codebase and noticed something this change. I’m not a maintainer or anything, so please don’t read this as a binding request.

File: util/builtin_operators.go (hunk 5 above)
Comment (hsluoyz) [reply util/builtin_operators.go:335]: @timconinx

Comment (timconinx) [reply util/builtin_operators.go:335]: @thetorpedodog 
I did change it, because the caching lead to concurrency problems in the `g` function. I discussed it in the related issue ([direct link](https://github.com/casbin/casbin/issues/1028#issuecomment-1153980413))

I'm not a real fan of replacing string bool maps by syncMaps, so if you see a better solution, then by all means

**Review by hsluoyz — APPROVED** [review]
